# **1 часть**

Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью polars

In [2]:
!pip install polars
import polars as pl

In [3]:
# Загрузка датасета
data = pl.read_csv('train.csv')

# Вывод первых 5 строк для проверки
print(data.head())

shape: (5, 12)
┌─────────────┬──────────┬────────┬───────────────────┬───┬───────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name              ┆ … ┆ Ticket    ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---               ┆   ┆ ---       ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str               ┆   ┆ str       ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪═══════════════════╪═══╪═══════════╪═════════╪═══════╪══════════╡
│ 1           ┆ 0        ┆ 3      ┆ Braund, Mr. Owen  ┆ … ┆ A/5 21171 ┆ 7.25    ┆ null  ┆ S        │
│             ┆          ┆        ┆ Harris            ┆   ┆           ┆         ┆       ┆          │
│ 2           ┆ 1        ┆ 1      ┆ Cumings, Mrs.     ┆ … ┆ PC 17599  ┆ 71.2833 ┆ C85   ┆ C        │
│             ┆          ┆        ┆ John Bradley (Fl… ┆   ┆           ┆         ┆       ┆          │
│ 3           ┆ 1        ┆ 3      ┆ Heikkinen, Miss.  ┆ … ┆ STON/O2.  ┆ 7.92

Выведите основную информацию о датасете: информацию о типах данных, число пропусков, средние значения и т.д.
hint: используйте методы и атрибуты датафрейма: describe(), dtypes, null_count(), mean()...

In [4]:
# Типы данных в колонках
print("Типы данных в колонках:")
print(data.dtypes)

Типы данных в колонках:
[Int64, Int64, Int64, String, String, Float64, Int64, Int64, String, Float64, String, String]


In [5]:
# Основная статистика по числовым колонкам
print("\nОсновные статистические показатели:")
print(data.describe())


Основные статистические показатели:
shape: (9, 13)
┌────────────┬─────────────┬──────────┬──────────┬───┬───────────┬───────────┬───────┬──────────┐
│ statistic  ┆ PassengerId ┆ Survived ┆ Pclass   ┆ … ┆ Ticket    ┆ Fare      ┆ Cabin ┆ Embarked │
│ ---        ┆ ---         ┆ ---      ┆ ---      ┆   ┆ ---       ┆ ---       ┆ ---   ┆ ---      │
│ str        ┆ f64         ┆ f64      ┆ f64      ┆   ┆ str       ┆ f64       ┆ str   ┆ str      │
╞════════════╪═════════════╪══════════╪══════════╪═══╪═══════════╪═══════════╪═══════╪══════════╡
│ count      ┆ 891.0       ┆ 891.0    ┆ 891.0    ┆ … ┆ 891       ┆ 891.0     ┆ 204   ┆ 889      │
│ null_count ┆ 0.0         ┆ 0.0      ┆ 0.0      ┆ … ┆ 0         ┆ 0.0       ┆ 687   ┆ 2        │
│ mean       ┆ 446.0       ┆ 0.383838 ┆ 2.308642 ┆ … ┆ null      ┆ 32.204208 ┆ null  ┆ null     │
│ std        ┆ 257.353842  ┆ 0.486592 ┆ 0.836071 ┆ … ┆ null      ┆ 49.693429 ┆ null  ┆ null     │
│ min        ┆ 1.0         ┆ 0.0      ┆ 1.0      ┆ … ┆ 110152    ┆

In [6]:
# Информация о пропусках
print("\nКоличество пропусков в каждой колонке:")
print(data.null_count())


Количество пропусков в каждой колонке:
shape: (1, 12)
┌─────────────┬──────────┬────────┬──────┬───┬────────┬──────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name ┆ … ┆ Ticket ┆ Fare ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---  ┆   ┆ ---    ┆ ---  ┆ ---   ┆ ---      │
│ u32         ┆ u32      ┆ u32    ┆ u32  ┆   ┆ u32    ┆ u32  ┆ u32   ┆ u32      │
╞═════════════╪══════════╪════════╪══════╪═══╪════════╪══════╪═══════╪══════════╡
│ 0           ┆ 0        ┆ 0      ┆ 0    ┆ … ┆ 0      ┆ 0    ┆ 687   ┆ 2        │
└─────────────┴──────────┴────────┴──────┴───┴────────┴──────┴───────┴──────────┘


In [7]:
# Средние значения (mean) для нужных колонок
print("\nСредние значения по некоторым ключевым колонкам:")
print(data.select(pl.mean('Age'), pl.mean('Fare')))


Средние значения по некоторым ключевым колонкам:
shape: (1, 2)
┌───────────┬───────────┐
│ Age       ┆ Fare      │
│ ---       ┆ ---       │
│ f64       ┆ f64       │
╞═══════════╪═══════════╡
│ 29.699118 ┆ 32.204208 │
└───────────┴───────────┘


Посчитайте количество пассажиров каждого класса (Pclass)
hint: используйте методы: get_column(), value_counts()...

In [8]:
# Получаем количество пассажиров каждого класса
passengers_by_class = data.get_column('Pclass').value_counts()

In [9]:
# Выводим результат
print(passengers_by_class)

shape: (3, 2)
┌────────┬───────┐
│ Pclass ┆ count │
│ ---    ┆ ---   │
│ i64    ┆ u32   │
╞════════╪═══════╡
│ 3      ┆ 491   │
│ 1      ┆ 216   │
│ 2      ┆ 184   │
└────────┴───────┘


Выведите количество выживших мужчин и женщин на корабле
hint: используйте методы: groupby(), agg(), pl.col(), sum()...

In [11]:
# Группируем по полу и считаем количество выживших
survivors_by_sex = data.group_by('Sex').agg([
    pl.sum('Survived').alias('Количество выживших')
])

In [12]:
# Выводим результат
print(survivors_by_sex)

shape: (2, 2)
┌────────┬─────────────────────┐
│ Sex    ┆ Количество выживших │
│ ---    ┆ ---                 │
│ str    ┆ i64                 │
╞════════╪═════════════════════╡
│ male   ┆ 109                 │
│ female ┆ 233                 │
└────────┴─────────────────────┘


Выведите часть таблицы с пассажирами, возраст которых больше 44 лет
hint: используйте методы: filter(), pl.col()...

In [13]:
# Фильтрация по возрасту > 44 лет
old_passengers = data.filter(pl.col('Age') > 44)

In [14]:
# Выводим результат
print(old_passengers)

shape: (115, 12)
┌─────────────┬──────────┬────────┬──────────────────┬───┬────────────┬─────────┬───────┬──────────┐
│ PassengerId ┆ Survived ┆ Pclass ┆ Name             ┆ … ┆ Ticket     ┆ Fare    ┆ Cabin ┆ Embarked │
│ ---         ┆ ---      ┆ ---    ┆ ---              ┆   ┆ ---        ┆ ---     ┆ ---   ┆ ---      │
│ i64         ┆ i64      ┆ i64    ┆ str              ┆   ┆ str        ┆ f64     ┆ str   ┆ str      │
╞═════════════╪══════════╪════════╪══════════════════╪═══╪════════════╪═════════╪═══════╪══════════╡
│ 7           ┆ 0        ┆ 1      ┆ McCarthy, Mr.    ┆ … ┆ 17463      ┆ 51.8625 ┆ E46   ┆ S        │
│             ┆          ┆        ┆ Timothy J        ┆   ┆            ┆         ┆       ┆          │
│ 12          ┆ 1        ┆ 1      ┆ Bonnell, Miss.   ┆ … ┆ 113783     ┆ 26.55   ┆ C103  ┆ S        │
│             ┆          ┆        ┆ Elizabeth        ┆   ┆            ┆         ┆       ┆          │
│ 16          ┆ 1        ┆ 2      ┆ Hewlett, Mrs.    ┆ … ┆ 248706     ┆ 16

# **2 часть**

Считайте датасет из файла train.csv (это данные о выживаемости на Титанике) с помощью pandas

In [18]:
import pandas as pd
import numpy as np
import bottleneck as bn

In [16]:
# Загрузка датасета
data = pd.read_csv('train.csv')

In [17]:
# Просмотрим первые несколько строк
print(data.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S  


Посчитайте средний возраст пассажиров и его стандартное отклонение с помощью bottleneck
hint: используйте методы: bn.nanmean(), bn.nanstd()

In [19]:
# Преобразуем колонку "Age" в Numpy-массив
ages_array = data['Age'].to_numpy()

In [20]:
# Рассчитываем средний возраст и стандартное отклонение с помощью bottleneck
average_age = bn.nanmean(ages_array)
age_std_deviation = bn.nanstd(ages_array)

In [21]:
# Выводим результаты
print(f"Средний возраст пассажиров: {average_age:.2f}")
print(f"Стандартное отклонение возраста: {age_std_deviation:.2f}")

Средний возраст пассажиров: 29.70
Стандартное отклонение возраста: 14.52


Для каждого пассажира умножьте значение столбца Fare на 1.3 и сохраните результаты как новый столбец Fare_new
hint: используйте методы: .itertuples() или apply()

In [22]:
# Умножаем значения в столбце Fare на 1.3 и создаем новый столбец Fare_new
data['Fare_new'] = data['Fare'] * 1.3

In [23]:
# Выводим обновленный датасет
print(data.head())

   PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  Fare_new  
0      0         A/5 21171   7.2500   NaN        S   9.42500  
1      0          PC 17599  71.2833   C85        C  92.66829  
2      0  STON/O2. 3101282   7.9250   NaN        S  10.30250  
3      0            113803  53.1000  C123        S  69.03000  
4    

# **3 часть**

Считайте датасет из файла Housing.csv (это данные о ценах домов) с помощью pandas

In [37]:
# Загрузка датасета
data = pd.read_csv('Housing.csv')

In [38]:
# Просмотрите первые несколько строк датасета
print(data.head())

      price  area  bedrooms  bathrooms  stories mainroad guestroom basement  \
0  13300000  7420         4          2        3      yes        no       no   
1  12250000  8960         4          4        4      yes        no       no   
2  12250000  9960         3          2        2      yes        no      yes   
3  12215000  7500         4          2        2      yes        no      yes   
4  11410000  7420         4          1        2      yes       yes      yes   

  hotwaterheating airconditioning  parking prefarea furnishingstatus  
0              no             yes        2      yes        furnished  
1              no             yes        3       no        furnished  
2              no              no        2      yes   semi-furnished  
3              no             yes        3      yes        furnished  
4              no             yes        2       no        furnished  


Для каждого столбца определите оптимальный с точки зрения потребления памяти тип данных - напишите свои выводы в комментариях

Поменяйте типы данных столбцов датафрейма на выбранные вами в прошлом пункте и сравните потребление памяти до и после оптимизации

In [39]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   price             545 non-null    int64 
 1   area              545 non-null    int64 
 2   bedrooms          545 non-null    int64 
 3   bathrooms         545 non-null    int64 
 4   stories           545 non-null    int64 
 5   mainroad          545 non-null    object
 6   guestroom         545 non-null    object
 7   basement          545 non-null    object
 8   hotwaterheating   545 non-null    object
 9   airconditioning   545 non-null    object
 10  parking           545 non-null    int64 
 11  prefarea          545 non-null    object
 12  furnishingstatus  545 non-null    object
dtypes: int64(6), object(7)
memory usage: 55.5+ KB
None


* Price: Цена недвижимости (тип данных — числовой, большой диапазон значений).
* Area: Площадь недвижимости (целочисленная переменная).
* Bedrooms: Количество спален (маленькие целые числа).
* Bathrooms: Количество санузлов (малые целые числа).
* Stories: Количество этажей (маленькие целые числа).
* Mainroad: Близость к главной дороге (строковый тип, бинарная переменная).
* Guestroom: Присутствие гостиной комнаты (строковый тип, бинарная переменная).
* Basement: Есть ли подвал (строковый тип, бинарная переменная).
* Hotwaterheating: Горячая вода централизованно (строковый тип, бинарная переменная).
* Airconditioning: Кондиционирование воздуха (строковый тип, бинарная переменная).
* Parking: Количество парковочных мест (целочисленное).
* Prefarea: Предпочтительная область (строковый тип, вероятно категориальный).
* Furnishingstatus: Оснащенность мебелью (строковый тип, категориальный).

* Числовые столбцы:
  * Price: Цены большие, но вряд ли нужны сверхточные вычисления. Тип данных можно оставить int64 или перейти на int32.
  * Area: Целочисленные площади домов, небольшой диапазон. Подходит тип int32 или даже int16.
  * Bedrooms, Bathrooms, Stories, Parking: Маленькие целые числа, подходят типы int8 или uint8 (unsigned integer).
* Категориальные столбцы:
  * Mainroad, Guestroom, Basement, Hotwaterheating, Airconditioning: Все эти столбцы бинарные ("yes/no"). Их лучше превратить в тип bool или, если сохранять оригинальную форму, в тип category.
  * Prefarea:  категориальная переменная. Ее можно привести к типу category.
  * Furnishingstatus:  категориальная переменная, тип данных можно поменять на category.

In [40]:
# Проверка текущего расхода памяти
print("Первоначальное потребление памяти:", data.memory_usage(deep=True).sum() / 1024, "KB")

Первоначальное потребление памяти: 251.72265625 KB


In [41]:
# Преобразование числовых столбцов
data['price'] = data['price'].astype('int32')
data['area'] = data['area'].astype('int32')
data['bedrooms'] = data['bedrooms'].astype('int8')
data['bathrooms'] = data['bathrooms'].astype('int8')
data['stories'] = data['stories'].astype('int8')
data['parking'] = data['parking'].astype('int8')

In [42]:
# Преобразование категориальных и бинарных столбцов
data['mainroad'] = data['mainroad'].astype('category')
data['guestroom'] = data['guestroom'].astype('category')
data['basement'] = data['basement'].astype('category')
data['hotwaterheating'] = data['hotwaterheating'].astype('category')
data['airconditioning'] = data['airconditioning'].astype('category')
data['prefarea'] = data['prefarea'].astype('category')
data['furnishingstatus'] = data['furnishingstatus'].astype('category')

In [43]:
# Проверка новой структуры
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 545 entries, 0 to 544
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   price             545 non-null    int32   
 1   area              545 non-null    int32   
 2   bedrooms          545 non-null    int8    
 3   bathrooms         545 non-null    int8    
 4   stories           545 non-null    int8    
 5   mainroad          545 non-null    category
 6   guestroom         545 non-null    category
 7   basement          545 non-null    category
 8   hotwaterheating   545 non-null    category
 9   airconditioning   545 non-null    category
 10  parking           545 non-null    int8    
 11  prefarea          545 non-null    category
 12  furnishingstatus  545 non-null    category
dtypes: category(7), int32(2), int8(4)
memory usage: 11.1 KB
None


In [44]:
# Повторно проверяем потребление памяти
print("Потребление памяти после оптимизации:", data.memory_usage(deep=True).sum() / 1024, "KB")

Потребление памяти после оптимизации: 11.876953125 KB


* Сокращение памяти за счет перехода к более эффективным типам данных.
* Увеличение производительности последующих операций.
* Удобная и быстрая подготовка данных для моделей машинного обучения и аналитики.

*Первоначальное потребление памяти: 251.72265625 KB*

*Потребление памяти после оптимизации: 11.876953125 KB*

Замена типов данных позволила добиться значительной экономии памяти, что улучшило удобство работы с датафреймом и повысило производительность будущих операций.